### PR TITLE
Add support for the transactional API

### PR DESCRIPTION
**Pull request by swastik**
This PR adds support for the Customer.io Transactional API.

### Changes

- Extracted a `BaseClient` class from the existing `Client` that contains all the request-related logic
- Updated `Client` to use the `BaseClient` class (and updated tests too)
- Added an `API` class to connect to the Transactional API and a `#send_email` method to send emails
- Added documentation

### CODE REVIEW

File: lib/customerio/base_client.rb [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+  DEFAULT_TIMEOUT  = 10
+
+  class BaseClient
+    def initialize(auth, options = {})
+      @auth = auth
+      @timeout = options[:timeout] || DEFAULT_TIMEOUT
+
+      @json = options.has_key?(:json) ? options[:json] : true
+      @base_uri = options[:base_uri] || DEFAULT_BASE_URI
+      @timeout = options[:timeout] || DEFAULT_TIMEOUT
```
Comment (jatinn) [inline lib/customerio/base_client.rb:14]: `@timeout` declared twice

File: lib/customerio/base_client.rb [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+module Customerio
+  DEFAULT_TIMEOUT  = 10
+
+  class BaseClient
+    def initialize(auth, options = {})
+      @auth = auth
+      @timeout = options[:timeout] || DEFAULT_TIMEOUT
+
+      @json = options.has_key?(:json) ? options[:json] : true
+      @base_uri = options[:base_uri] || DEFAULT_BASE_URI
```
Comment (jatinn) [inline lib/customerio/base_client.rb:13]: whats `DEFAULT_BASE_URI` in this case? feels like maybe it should be an error if missing

File: lib/customerio/base_client.rb [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+      @timeout = options[:timeout] || DEFAULT_TIMEOUT
+    end
+
+    def request(method, path, body = nil, headers = {})
+      execute(method, path, body, headers)
+    end
+
+    private
+
+    def extract_attributes(args)
```
Comment (jatinn) [inline lib/customerio/base_client.rb:23]: only used by `track` method and is duplicated in `Client` class as well, can likely clean up one instance.

File: lib/customerio/api.rb [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  class ApiClient
+    DEFAULT_BASE_URI = 'https://api.customer.io'
+
+    def initialize(app_key, options = {})
+      options[:base_uri] = DEFAULT_BASE_URI if options[:base_uri].nil? || options[:base_uri].empty?
+      @client = Customerio::BaseClient.new({ app_key: app_key }, options)
+    end
+
+    def send_email(payload)
+      @client.request(:post, send_email_path, payload)
```
Comment (jatinn) [inline lib/customerio/api.rb:14]: track endpoints wrap calls with `verify_response` any particular reason to not perform that check here?

**Review by jatinn — CHANGES_REQUESTED** [review]

File: lib/customerio/api.rb (hunk 4 above)
Comment (swastik) [reply lib/customerio/api.rb]: Good point! No reason to not do it, so I've moved `verify_response` into the base client and I'm now calling that method here.

**Review by jatinn — APPROVED** [review]

File: lib/customerio/base_client.rb [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        raise InvalidRequest.new("Invalid request method #{method.inspect}")
+      end
+    end
+
+    def add_request_body(req, body)
+      if @json
+        req.add_field('Content-Type', 'application/json')
+        req.body = MultiJson.dump(body)
+      else
+        req.add_field('Content-Type', 'application/x-www-form-urlencoded')
```
Comment (newhook) [inline lib/customerio/base_client.rb:79]: What is this case for? We don't support this with go. I don't think we should support it with ruby.

File: lib/customerio/client.rb [hunk 6]
```diff
@@ -48,44 +37,49 @@ def track(*args)
       if args.length == 1
         # Only passed in an event name, create an anonymous event
         event_name = args.first
+        raise ParamError.new("event_name must be a non-empty string") if is_empty?(event_name)
```
Comment (newhook) [inline lib/customerio/client.rb:40]: This isn't what the golang library does.

File: lib/customerio/client.rb [hunk 7]
```diff
@@ -48,44 +37,49 @@ def track(*args)
...
-      raise ParamError.new("customer_id must be a non-empty string") unless customer_id != "" and !customer_id.nil?
-      raise ParamError.new("device_id must be a non-empty string") unless device_id != "" and !device_id.nil?
+      raise ParamError.new("customer_id must be a non-empty string") if is_empty?(customer_id)
+      raise ParamError.new("device_id must be a non-empty string") if is_empty?(device_id)
       
-      verify_response(request(:delete, device_id_path(customer_id, device_id)))
+      @client.request_and_verify_response(:delete, device_id_path(customer_id, device_id))
     end
 
     def add_to_segment(segment_id, customer_ids)
```
Comment (newhook) [inline lib/customerio/client.rb:85]: golang library checks that the segment_id is not empty.

File: lib/customerio/client.rb (hunk 6 above)
Comment (swastik) [reply lib/customerio/client.rb:40]: Dropped the case that handles single args and sends an anonymous event.

**Review by jatinn — APPROVED** [review]

**Comment by kylefox** [comment]
> **Removed**
> `add_to_segment` and `remove_from_segment` methods

Is there any guidance on how we're supposed to migrate our integration from these methods to their new equivalent?

The API documentation simply states the [`add_to_segment`](https://customer.io/docs/api/#operation/add_to_segment) and [`remove_from_segment`](https://customer.io/docs/api/#operation/remove_from_segment) methods are deprecated, but offers no information on how to add customers to segments with the new API, from what I can tell.

**Comment by rlisowski** [comment]
@kylefox I'm looking into the same thing, and I think you achieve it by making a data-driven segment and simply sending an event https://customer.io/docs/segments

**Pull request merged**